File: tasks/discover-start-pages.rb

require 'json'
require 'open-uri'


existing_services = []

Dir.glob('app/services/*.json') do |file|
  next if file.end_with?('_template.json')

  json = JSON.parse(File.read(file))
  json['file'] = file
  existing_services << json
end


existing_live_service_hosts = existing_services.collect do |existing_service|
  if existing_service['liveservice'].nil?
    nil
  else
    URI.parse(existing_service['liveservice']).host
  end
end.compact

count = 1000
start = 0
results = []
service_domains = []

ignored_start_pages = File.read("tasks/ignored_start_page_urls.txt").split("\n")

file = URI.open("https://www.gov.uk/api/search.json?filter_format=transaction&count=#{count}&start=#{start}")

print "."
json = JSON.parse(file.read)

results = json['results']

start_pages = results.collect {|x| "https://www.gov.uk" + x['link'] } - ignored_start_pages

missing_start_pages = start_pages.select do |start_page|

  match = existing_services.detect {|service| Array(service['start-page']).include?(start_page) }

  !match
end

puts "#{missing_start_pages.size} missing start pages: "
puts missing_start_pages.sort

File: tasks/discover-start-pages.js
import ignoredStartPages from "../data/ignored-start-pages.json" with { type: "json" };
import { getServices } from "../lib/data.js";
import { getTransactionResults } from "../lib/search-api.js";

const services = await getServices();
const existingStartPages = [];
const newStartPages = [];

for (const service of services) {
  const startPages = service.startPage;

  if (Array.isArray(startPages)) {
    existingStartPages.push(...startPages);
  } else if (startPages) {
    existingStartPages.push(startPages);
  }
}

try {
  const results = await getTransactionResults();

  const startPages = results.map(
    (result) => `https://www.gov.uk${result.link}`,
  );

  for (const startPage of startPages) {
    if (
      !existingStartPages.includes(startPage) &&
      !ignoredStartPages.includes(startPage)
    ) {
      newStartPages.push(startPage);
    }
  }

  console.log(`${newStartPages.length} new start pages found:\n`);
  for (const startPage of newStartPages.sort()) {
    console.log(startPage);
  }
} catch (error) {
  console.error(error.message);
}
